### PR TITLE
Support `with_array_context` for DOFArrays

### PR DESCRIPTION
**Pull request by kaushikcfd**
See https://github.com/inducer/arraycontext/pull/158#issuecomment-1120080728 for context.

- :warning: Needs to go in conjunction with https://github.com/inducer/arraycontext/pull/158
- [ ] Drop "Drop before merge" commit.

### CODE REVIEW

File: meshmode/dof_array.py [hunk 1]
```diff
@@ -378,25 +378,9 @@ def _raise_index_inconsistency(i, stream_i):
...
-def _thaw_dofarray(ary, actx):
-    if ary.array_context is not None:
-        raise ValueError("cannot thaw DOFArray that already has an array context")
-
-    return type(ary)(
-        actx,
-        tuple(actx.thaw(subary) for subary in ary._data))
+@with_array_context.register(DOFArray)
+def _with_actx_dofarray(ary, actx):
+    return type(ary)(actx, ary._data)
```
Comment (alexfikl) [inline meshmode/dof_array.py:383]: Add a check that the arrays in `_data` are in `actx.array_types`? Maybe under an `if __debug__`

Comment (kaushikcfd) [reply meshmode/dof_array.py:383]: Sure, recursively in each component of `_data`?

Comment (alexfikl) [reply meshmode/dof_array.py:383]: That shouldn't be necessary, since `DOFArray._data` should just contain the base array type (right?).

Comment (kaushikcfd) [reply meshmode/dof_array.py:385]: Done, thanks! (I thought DOFArrays could also contain numpy object arrays of base arrays).

**Comment by inducer** [comment]
Went ahead and force-merged this based on earlier CI passing.

**Pull request merged**